RST-1118 3D Orientation Constraint

ayrton04 opened the pull request.

ayrton04 committed:
Adding 3D orientation constraints

ayrton04 commented:
Will squash once approved.

svwilliams commented on fuse_constraints/include/fuse_constraints/absolute_orientation_3d_stamped_constraint.h:
```diff
@@ -0,0 +1,182 @@
+   *
+   * @param[in] orientation The variable representing the orientation components of the pose
+   * @param[in] mean        The measured/prior orientation as a quaternion (4x1 vector: w, x, y, z)
+   * @param[in] covariance  The measurement/prior covariance (3x3 matrix: roll, pitch, yaw)
```
Unless you are transforming the input somewhere, this is not a (roll, pitch, yaw) covariance; it is a (qx, qy, qz) covariance.

svwilliams commented on fuse_constraints/include/fuse_constraints/absolute_orientation_3d_stamped_constraint.h:
```diff
@@ -0,0 +1,182 @@
+  AbsoluteOrientation3DStampedConstraint(
+    const fuse_variables::Orientation3DStamped& orientation,
+    const Eigen::Vector4d& mean,
+    const Eigen::MatrixXd& covariance);
```
Should this be a fixed-size `Eigen::Matrix3d`?

ayrton04 replied:
Yeah, [it was originally](https://github.com/locusrobotics/fuse/pull/20/commits/84a68dae6e5551215845cc02c1db4e48357ff727#diff-e3ceed27358149a678861010cb692218R75), but then we tried to allow for selecting axes.

svwilliams commented on fuse_constraints/include/fuse_constraints/absolute_orientation_3d_stamped_constraint.h:
```diff
@@ -0,0 +1,182 @@
+   *
+   * Order is (qx, qy, qz)
+   */
+  const Eigen::MatrixXd& sqrtInformation() const { return sqrt_information_; }
```
Fixed-size matrix here too?

svwilliams commented on fuse_constraints/include/fuse_constraints/absolute_orientation_3d_stamped_euler_constraint.h:
```diff
@@ -0,0 +1,146 @@
+{
+
+/**
+ * @brief A constraint that represents either prior information about a 3D orientation, or a direct measurement of the
```
Probably mention something about the measurement/error being in Euler space instead of quaternion space

svwilliams commented on fuse_constraints/include/fuse_constraints/absolute_orientation_3d_stamped_euler_constraint.h:
```diff
@@ -0,0 +1,146 @@
+  /**
+   * @brief Read-only access to the measured/prior vector of mean values.
+   *
+   * Order is defined by the provided \p axes parameter. This mean() function deviates from all other
```
I dislike this constraint working differently than the others. And don't worry, I remember that I'm the one that wrote it. I'm just curious which implementation you think is "wrong"? This basically returns the same thing that was provided as input. That feels "more right" than what is being done in the other constraints. Just possibly less useful.

If we keep it this way, we may want to add read-only access to the `axes` vector. That way the mean/cov can be correctly interpreted by someone who didn't create it.

ayrton04 replied:
Agreed on all counts. I think it's more strange for the use to give it a mean and then get a different one back.

svwilliams commented on fuse_constraints/include/fuse_constraints/absolute_orientation_3d_stamped_euler_constraint.h:
```diff
@@ -0,0 +1,146 @@
+protected:
+  Eigen::VectorXd mean_;  //!< The measured/prior mean vector for this variable
+  Eigen::MatrixXd sqrt_information_;  //!< The square root information matrix
+  std::vector<Euler> axes_;
```
Doxygen description?

svwilliams commented on fuse_constraints/src/absolute_orientation_3d_stamped_euler_constraint.cpp:
```diff
@@ -0,0 +1,82 @@
+    axes_(axes)
+{
+  assert(covariance.rows() == static_cast<int>(axes.size()));
+  assert(covariance.cols() == static_cast<int>(axes.size()));
```
Add asserts for the mean vector...as long as we are having asserts

svwilliams commented on fuse_constraints/src/absolute_orientation_3d_stamped_euler_constraint.cpp:
```diff
@@ -0,0 +1,82 @@
+    new NormalPriorOrientation3DEulerCostFunctor(sqrt_information_, mean_, axes_), axes_.size());
+}
+
+}  // namespace fuse_constraints
```
Missing newline

svwilliams commented on fuse_variables/include/fuse_variables/orientation_3d_stamped.h:
```diff
@@ -91,15 +93,49 @@ class Orientation3DStamped final : public FixedSizeVariable<4>, public Stamped
    */
   explicit Orientation3DStamped(const ros::Time& stamp, const fuse_core::UUID& device_id = fuse_core::uuid::NIL);
 
+  template <typename T>
```
Should these be moved to util? Doesn't really matter; just curious where you think the better home would be.

ayrton04 replied:
Sure. It seems reasonable that someone somewhere might want this functionality outside of this variable.

ayrton04 replied:
We'd have to make a `util.h` inside `fuse_variables`, though, because I use these methods inside a variable type, and if we depend on `fuse_constraints`, where the current `util.h` lives, it'll be a circular dependency.

svwilliams commented on fuse_constraints/include/fuse_constraints/util.h:
```diff
@@ -59,6 +59,14 @@ void wrapAngle2D(T& angle)
+template<typename T>
+T wrapAngle2D(const T& angle)
+{
+  T wrapped = angle;
```
`T wrapped(angle);`
`T wrapped = T(angle);`
???
Trying to decide how our policy of explicit casting should be implemented here. Or even if any of these options are actually different.

ayrton04 replied:
`T wrapped { angle };` may be better, since it doesn't permit narrowing.

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_cost_functor.h:
```diff
@@ -0,0 +1,139 @@
+   * @param[in] b The orientation measurement or prior in order (w, x, y, z)
+   */
+  NormalPriorOrientation3DCostFunctor(
+    const Eigen::MatrixXd& A,
```
Fixed size 3?

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_cost_functor.h:
```diff
@@ -0,0 +1,139 @@
+ * The cost function is of the form:
+ *
+ *             ||    [  x - b(1)] ||^2
+ *   cost(x) = ||A * [  y - b(2)] ||
```
This is really something more like: `cost(x) = || A * imag( q^-1 * b) ||^2`  ... or something. I don't know my quaternion math off the top of my head. But the description should be updated with the correct math.

ayrton04 replied:
```
 *             ||    [  (b * q^-1)(1)  ] ||^2
 *   cost(x) = ||A * [  (b * q^-1)(2)  ] ||
 *             ||    [  (b * q^-1)(3)  ] ||
```

Is that clear?

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_cost_functor.h:
```diff
@@ -0,0 +1,139 @@
+/**
+ * @brief Create a prior cost function on a 3D orientation variable (quaternion)
+ *
+ * The Ceres::NormalPrior cost function only supports a single variable. This is a convenience cost function that
```
Just kill these two sentences. They are basically meaningless here. We are not using the Ceres::NormalPrior, nor is there any good way to implement this cost function using the Ceres::NormalPrior even if you wanted to.

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -0,0 +1,149 @@
+
+}  // namespace fuse_constraints
+
+#endif  // FUSE_CONSTRAINTS_NORMAL_PRIOR_ORIENTATION_3D_EULER_COST_FUNCTOR_H
```
Missing newline

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -0,0 +1,149 @@
+private:
+  Eigen::MatrixXd A_;  //!< The residual weighting matrix, most likely the square root information matrix
+  Eigen::VectorXd b_;  //!< The measured 3D orientation (quaternion) value
+  std::vector<Euler> axes_;  //!< ???
```
doxygen comments

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -0,0 +1,149 @@
+#include <fuse_constraints/util.h>
+#include <fuse_variables/orientation_3d_stamped.h>
+
+#include <angles/angles.h>
```
Doesn't seem like `angles` is used

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -0,0 +1,149 @@
+ * The cost function is of the form:
+ *
+ *             ||    [  roll - b(1)] ||^2
+ *   cost(x) = ||A * [  pitch - b(2)] ||
```
clean up spacing

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -0,0 +1,149 @@
+/**
+ * @brief Create a prior cost function on a 3D orientation variable (quaternion)
+ *
+ * The Ceres::NormalPrior cost function only supports a single variable. This is a convenience cost function that
```
Delete these two sentences

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -0,0 +1,149 @@
+{
+
+/**
+ * @brief Create a prior cost function on a 3D orientation variable (quaternion)
```
...using Euler roll, pitch, yaw measurements

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -0,0 +1,149 @@
+ *   cost(x) = ||A * [  pitch - b(2)] ||
+ *             ||    [  yaw - b(3)] ||
+ *
+ * where the matrix A and the vector b are fixed and (roll, pitch, yaw) are the components of the 3D orientation
```
...and (roll, pitch, yaw) are the Euler angles extracted from the 3D orientation variable.

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -0,0 +1,149 @@
+   * @param[in] A The residual weighting matrix, most likely the square root information matrix in order
+   *              (roll, pitch, yaw)
+   * @param[in] b The orientation measurement or prior in order (roll, pitch, yaw)
+   * @param[in] axes The Euler angle axes for which we want to compute errors. Defaults to all axes. 
```
There are no defaults listed

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -0,0 +1,149 @@
+   * 
+   * @param[in] A The residual weighting matrix, most likely the square root information matrix in order
+   *              (roll, pitch, yaw)
+   * @param[in] b The orientation measurement or prior in order (roll, pitch, yaw)
```
Order is controlled by the `axes` parameter

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -0,0 +1,149 @@
+  /**
+   * @brief Construct a cost function instance
+   * 
+   * @param[in] A The residual weighting matrix, most likely the square root information matrix in order
```
Order is controlled by the axes parameter

svwilliams reviewed (changes requested):
Mostly/entirely documentation changes

ayrton04 commented:
Addressed the feedback. Not sure if we should push all the `util` methods down into `fuse_variables` or even `fuse_core`, since we now have utility methods in both `fuse_constraints` and `fuse_variables`.

svwilliams commented on fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.h:
```diff
@@ -35,29 +35,28 @@
- *             ||    [  roll - b(1)] ||^2
- *   cost(x) = ||A * [  pitch - b(2)] ||
- *             ||    [  yaw - b(3)] ||
+ *             ||    [  roll  - b(1)  ] ||^2
```
Should probably zero-index the components. And it's a little deceiving because the order is defined by the `axes` parameter. I don't really have a suggestion on how to make it both readable _and_ not deceiving.

Maybe just by example?
```
The order of the measured axes is defined in the \p axes variable. For example, if the yaw and roll were measured in that order, then:
b_ = [ measured_yaw  ]
     [ measured_roll ]

cost(x) = || A * [ yaw(x)  - b_(0) ] ||^2
          ||     [ roll(x) - b_(1) ] ||

```

svwilliams reviewed (approved):
Works for me
